openedx/core/djangoapps/credit/signals.py

```diff
@@ -70,3 +82,209 @@ def listen_for_grade_calculation(sender, username, grade_summary, course_key, de
+    # parent are 'vertical' and 'sequential' respectively);
+    # otherwise add groups to current ICRV's parent only
+    access_groups_id_list = [VerificationPartitionScheme.NON_VERIFIED, VerificationPartitionScheme.VERIFIED_ALLOW]
+    access_dict = _access_dict(group_configuration, access_groups_id_list)
+    parent_block = block.get_parent()
+    grandparent_block = parent_block.get_parent()
+    ancestor_categories = [parent_block.location.category, grandparent_block.location.category]
+
+    # update 'group_access' field of parent only (immediate siblings
+    # will have access control)
+    ancestor_for_update = parent_block
+    if len(set(ancestor_categories).difference(set(GATED_COURSE_CATEGORIES))) == 0:
+        # category of parent and grand parent are 'vertical' and
+        # 'sequential' respectively so update 'group_access' field of
+        # grandparent
+        ancestor_for_update = grandparent_block
+
+    ancestor_for_update.group_access = access_dict
+    modulestore().update_item(ancestor_for_update, user_id)
+
+
+def _update_course_user_partitions(course_key, user_id, new_user_partitions):
+    """ Add provided user partitions in 'new_user_partitions' in
+    'user_partitions' field of provided course.
+
+    Remove all user partitions of course with same partition schemes as in
+    provided list 'new_user_partitions'.
+
+    Args:
+        course_key (CourseKey): Identifier for the course
+        user_id (Int): ID of user who has publish the course
+        new_user_partitions (List): List of user partitions
+
+    """
+    course = modulestore().get_course(course_key, depth=0)
+    # get unique set of partition schemes for update and filter course user
+    # partitions with any of the unique set of partition schemes
+    course_user_partitions_old = [user_partition.id for user_partition in course.user_partitions]
+    filter_partition_schemes = set([user_partition.scheme.name for user_partition in new_user_partitions])
+    course_user_partitions = filter_by_scheme(course.user_partitions, filter_partition_schemes)
+    # new add user partitions 'new_user_partitions' in filter course user partitions
+    course.user_partitions = course_user_partitions + new_user_partitions
+    # TODO: Check that course and modules are properly updated in database, preserve those user partition which scheme name not in 'new_user_partitions'.
+    modulestore().update_item(course, user_id)
+    course_user_partitions_new = [user_partition.id for user_partition in course.user_partitions]
+
+    # sync blocks with access roles according to new user partitions for course
+    deleted_partitions = list(set(course_user_partitions_old) - set(course_user_partitions_new))
+
+    # sync blocks with access roles according to new user partitions for the
+    # course so that user is not denied access for blocks with non-existing
+    # user partitions/groups
+    _sync_course_content_deleted_partitions(course_key, deleted_partitions)
```

Comment:
I'd need to think about this some more, but my first impression is that this seems reasonable.  If I'm understanding this correctly, we'd be adding:

* One read operation to retrieve all content that with `group_access` set.
* One bulk write operation to update blocks with deleted partitions (which in the most common case should be none).

One alternative could be to loosen the constraints on user partitions, so that partitions that aren't included in the course descriptor are simply ignored.  This would avoid the additional queries on publish, but would leave junk in Mongo.  @jimabramson what do you think?